### ISSUE TITLE
[EPIC] A collection of tickets for supporting WASM in DataFusion

### ISSUE THREAD

**Issue by alamb** [apache/datafusion#13815]
### Is your feature request related to a problem or challenge?

DataFusion can be compiled today to WASM with some care. However it is somewhat tricky and the experience could be improved. 

### Describe the solution you'd like

This ticket tracks various ways we could make the experience better

### Describe alternatives you've considered

- [ ] https://github.com/apache/datafusion/issues/7652

**Comment by alamb** [comment]
> I’m very interested in contributing to the WASM support for DataFusion project as part of GSoC 2025. Enhancing embeddability and ensuring robust WASM integration aligns with my passion for building reliable and maintainable software

Thank you @savaliyabhargav 

I think the key challenge is the same as in this ticket's description: 

> DataFusion can be compiled today to WASM with some care. However it is somewhat tricky and the experience could be improved.

I suggest some key milestones:
1. Write a blog post explaining what WASM is, why it is important, and how to use DataFusion with it
2. Add example / section to the documentation about compiling and building with WASM
3. Make several example WASM apps with DataFusion -- for example https://github.com/apache/datafusion/issues/13818
4. Stretch goal -- prototype how we would support WASM user defined functions

**Comment by matthewmturner** [comment]
> 4. Stretch goal -- prototype how we would support WASM user defined functions

On this point, I was able to get WASM UDFs working in [dft](https://github.com/datafusion-contrib/datafusion-dft/tree/main/crates/datafusion-udfs-wasm) and as part of the next release I was planning to publish that crate so others could use it. @savaliyabhargav I would be happy to collaborate on that if you are interested in it.

**Comment by savaliyabhargav** [comment]
@matthewmturner  yes sure i am interested can you please give me more detail about it

**Comment by matthewmturner** [comment]
For the WASM UDFs they just need some more real world testing / benchmarking.  To be honest, the other points @alamb mentioned would probably better benefit the DataFusion ecosystem at this stage.  But, if WASM UDFs are what interest you then you could make an issue on the [dft](https://github.com/datafusion-contrib/datafusion-dft) repo and we could discuss some tangible next steps for that.

**Comment by qstommyshu** [comment]
Hi, I'm also interested in working on this project for Gsoc 2025. I'm currently a master student at Georgia Institute of Technology, I had some experiences with WASM and Rust in my undergrad study, and I would like to see how that can be incorporated with DataFusion!

Are there any prerequisite for this project? I suppose the points below would be the key points for our proposal, right? 

> I suggest some key milestones:

> Write a blog post explaining what WASM is, why it is important, and how to use DataFusion with it
> Add example / section to the documentation about compiling and building with WASM
> Make several example WASM apps with DataFusion -- for example https://github.com/apache/datafusion/issues/13818
> Stretch goal -- prototype how we would support WASM user defined functions

I'm new to DataFusion but I've heard of the name for a long long time.  

1. Can you please give me some relative materials to look at for this project? or working on some "good first issue" should be enough for me to get familiar with DataFusion?
2. Can you please give me some guidance on how to work on an precise and appropriate proposal?

Thanks in advance!

**Comment by savaliyabhargav** [comment]
@matthewmturner Please provide me with some information on DFT so that I can learn and work on your project effectively

**Comment by qstommyshu** [comment]
Hi @alamb and @waynexia ,

I've drafted a [proposal](https://github.com/qstommyshu/datafusion-wasm) for this project for Gsoc 2025. Would you mind taking a looking and giving me some feedback? I plan to explore more on datafusion WASM related  items after finishing my current PR sets on migrating tests to `insta`.

Thanks in advance!

**Comment by logankaser** [comment]
`datafusion` `54` regresses pretty far on wasm support, with default features requiring zstd and thus full Emscripten to compile.
`53.1` is the last version that it's reasonably possible to use with wasm

**Comment by kylebarron** [comment]
Is it only the inclusion of the feature as default that changed? Did you try opting-in manually to specific features and excluding zstd?

FWIW in other projects (e.g. parquet) I've been able to compile zstd to `wasm32-unknown-unknown` without any issue. If you're on macos [you need to switch the clang you use to compile](https://github.com/kylebarron/parquet-wasm/blob/main/DEVELOP.md#macos).

**Comment by alamb** [comment]
> `datafusion` `54` regresses pretty far on wasm support, with default features requiring zstd and thus full Emscripten to compile. `53.1` is the last version that it's reasonably possible to use with wasm

@logankaser  can you please provide a reproducer for this?

We have a wasm test module https://github.com/apache/datafusion/tree/main/datafusion/wasmtest precisely designed to ensure DataFusion can be compiled under wasm. If it lacking coverage I think we should fix it.
